File: ASD/ordered_list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.prev = None
        self.next = None


class OrderedList:
    def __init__(self, asc):
        self.head = None
        self.tail = None
        self.__ascending = asc

    def compare(self, v1, v2):
        if v1 < v2:
            return -1
        elif v1 > v2:
            return 1
        else:
            return 0

    def add(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return
        curr_node = self.head
        prev = None
        while curr_node is not None and ((self.__ascending and self.compare(value, curr_node.value) == 1) or
                                         (not self.__ascending and self.compare(value, curr_node.value) == -1)):
            prev = curr_node
            curr_node = curr_node.next
        if prev is None:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        elif curr_node is None:
            prev.next = new_node
            new_node.prev = prev
            self.tail = new_node
        else:
            prev.next = new_node
            new_node.prev = curr_node.prev
            new_node.next = curr_node
            curr_node.prev = new_node

    def find(self, val):
        curr_node = self.head
        while curr_node is not None:
            comparison = self.compare(val, curr_node.value)
            if comparison == 0:
                return curr_node
            elif (self.__ascending and comparison == -1) or (not self.__ascending and comparison == 1):
                return None
            curr_node = curr_node.next
        return None

    def delete(self, val):
        curr_node = self.head
        while curr_node is not None:
            if curr_node.value == val:
                if curr_node.prev is not None:
                    curr_node.prev.next = curr_node.next
                else:
                    self.head = curr_node.next
                if curr_node.next is not None:
                    curr_node.next.prev = curr_node.prev
                else:
                    self.tail = curr_node.prev
                return
            curr_node = curr_node.next
        return

    def clean(self, asc):
        self.__ascending = asc
        self.head = None
        self.tail = None

    def len(self):
        count = 0
        curr_node = self.head
        while curr_node is not None:
            count += 1
            curr_node = curr_node.next
        return count

    def get_all(self):
        r = []
        node = self.head
        while node != None:
            r.append(node)
            node = node.next
        return r


class OrderedStringList(OrderedList):
    def __init__(self, asc):
        super(OrderedStringList, self).__init__(asc)

    def compare(self, v1, v2):
        v1_s = v1.strip()
        v2_s = v2.strip()
        if v1_s < v2_s:
            return -1
        elif v1_s == v2_s:
            return 0
        else:
            return 1

File: java_oop/OrderedList.java
// 7. Упорядоченный список
import java.util.*;

class Node<T>
{
    public T value;
    public Node<T> next, prev;

    public Node(T _value)
    {
        value = _value;
        next = null;
        prev = null;
    }
}

public class OrderedList<T>
{
    public Node<T> head, tail;
    private boolean _ascending;

    public OrderedList(boolean asc)
    {
        head = null;
        tail = null;
        _ascending = asc;
    }

    // Метод сравнения двух значений
    public int compare(T v1, T v2)
    {
        // Сложность O(1)

        if (v1 == null && v2 == null) return 0;
        if (v1 == null) return -1;
        if (v2 == null) return 1;

        if (v1 instanceof Number && v2 instanceof Number)
        {
            double a = ((Number)v1).doubleValue();
            double b = ((Number)v2).doubleValue();
            if (a < b) return -1;
            if (a > b) return 1;
            return 0;
        }

        if (v1 instanceof String && v2 instanceof String)
        {
            String a = ((String)v1).trim();
            String b = ((String)v2).trim();
            int c = a.compareTo(b);
            if (c < 0) return -1;
            if (c > 0) return 1;
            return 0;
        }

        if (v1 instanceof Comparable && v2 instanceof Comparable)
        {
            int c = ((Comparable)v1).compareTo(v2);
            if (c < 0) return -1;
            if (c > 0) return 1;
            return 0;
        }

        return 0;
    }

    // Добавление нового элемента по значению
    public void add(T value)
    {
        // Сложность O(n)
        Node<T> n = new Node<T>(value);

        if (head == null)
        {
            head = n;
            tail = n;
            return;
        }

        Node<T> cur = head;

        while (cur != null)
        {
            int cmp = compare(value, cur.value);
            boolean insertBefore = _ascending ? (cmp <= 0) : (cmp >= 0);

            if (insertBefore)
            {
                n.next = cur;
                n.prev = cur.prev;

                if (cur.prev != null)
                    cur.prev.next = n;
                else
                    head = n;

                cur.prev = n;
                return;
            }

            cur = cur.next;
        }

        n.prev = tail;
        tail.next = n;
        tail = n;
    }

    public Node<T> find(T val)
    {
        // Сложность O(n)
        Node<T> cur = head;

        while (cur != null)
        {
            int cmp = compare(cur.value, val);

            if (cmp == 0)
                return cur;

            if (_ascending)
            {
                if (cmp > 0) return null;
            }
            else
            {
                if (cmp < 0) return null;
            }

            cur = cur.next;
        }

        return null;
    }

    // Удаление самого первого (одного) элемента по значению: Delete()
    public void delete(T val)
    {
        // Сложность O(n)
        Node<T> node = find(val);
        if (node == null) return;

        if (node.prev != null)
            node.prev.next = node.next;
        else
            head = node.next;

        if (node.next != null)
            node.next.prev = node.prev;
        else
            tail = node.prev;
    }

    public void clear(boolean asc)
    {
        // Сложность O(1)
        _ascending = asc;
        head = null;
        tail = null;
    }

    public int count()
    {
        // Сложность O(n)
        int c = 0;
        Node<T> cur = head;

        while (cur != null)
        {
            c++;
            cur = cur.next;
        }

        return c;
    }

    ArrayList<Node<T>> getAll()
    {
        // Сложность O(n)
        ArrayList<Node<T>> r = new ArrayList<Node<T>>();
        Node<T> node = head;
        while (node != null)
        {
            r.add(node);
            node = node.next;
        }
        return r;
    }
}

// Рефлексия
// В рекомендациях указаны оптимальные алгоримы, как и должно все работать.
// Полезно знать как оно работает на уровень ниже, при вызове метода особо об этом не задумываешься.